### PR TITLE
INT-4565: Fix IntComponentScan for profiles

### PR DESCRIPTION
**Pull request by artembilan**
JIRA: https://jira.spring.io/browse/INT-4565

* Propagate an `Environment` to the internal `ClassPathScanningCandidateComponentProvider`
in the `IntegrationComponentScanRegistrar` for proper profiles activation
* Ensure the logic works in the `GatewayInterfaceTests`
* Some polishing and performance improvement for the `GatewayInterfaceTests`
* Add a note about `@Profile` in the `gateway.adoc`
* Polishing for the `gateway.adoc`

**Cherry-pick to 5.0.x & 4.3.x**

### CODE REVIEW

**Review by garyrussell — CHANGES_REQUESTED** [review]

File: spring-integration-core/src/test/java/org/springframework/integration/gateway/GatewayInterfaceTests.java [hunk 1]
```diff
@@ -652,6 +662,7 @@ public GatewayProxyFactoryBean annotationGatewayProxyFactoryBean() {
 
 	@MessagingGateway
 	@TestMessagingGateway
+	@Profile("gatewayTest")
```
[the remark this answers is not in the document]
Comment (artembilan) [reply spring-integration-core/src/test/java/org/springframework/integration/gateway/GatewayInterfaceTests.java]: Added `@Profile("notActiveProfile")` gateway

**Comment by garyrussell** [comment]
Checkstyle
```
[ant:checkstyle] [ERROR] /home/travis/build/spring-projects/spring-integration/spring-integration-core/src/test/java/org/springframework/integration/gateway/GatewayInterfaceTests.java:383:104: ',' is not followed by whitespace. [WhitespaceAfter]
```

**Review by garyrussell — APPROVED** [review]

**Comment by garyrussell** [comment]
Lots of conflicts on cherry-pick

**Comment by artembilan** [comment]
Cherry-picked to `5.0.x` as https://github.com/spring-projects/spring-integration/commit/15024858d465423ac0735963b1493b0e2788acec after fixing conflicts.

**Pull request merged**